### DISCUSSION TITLE
Issues with both commanders while combining with Optitrack data

### DISCUSSION THREAD

**Discussion by SteveChex** [bitcraze/discussions#399] [Q&A]
Hi,

I'm currently programming my crazyflie to work with an Optitrack system and the Python lib. Using as reference the Qualisis example, I was able to send only the position data from the mocap to the drone, and it is able to estimate it's own position with low error.  But now, when I try to hover it using the commander, it doesn't work. 
More exactly, I was trying with the zdistance, hover, position and setpoint functions, and only with the zdistance and setpoint the drone was able to take off, but it does not handle the horizontal position. Also I have tried to change the flightmode parameters poshold, althold and posSet to keep it on position, but it does not change anything. With the hover function, the drone isn't able to take off, and with the position function, the drone take off to a close position to the expected, but after a few it start to tilt and speed until it crashes to the neaest wall. 

I also tried the High level commander, just in case someting was wrong with the first. I used the Autonomus_Sequence_High_level.py example and modified the trajectory to a straight line, just to test. When the trajectory is relative, it doesn't take off, and when it's not, the drone takes off, but making little quick jumps and never reaching the expected point. 

I have to mention that I worked with the Thread, Event and Keyboard libraries. 

I don't know if this is for sending an incorrect setpoint or a bad configuration, but I would apreciate any help.

There is a fragment of my main code, as reference. It is based on the Qualisis Mocap example. 
The "RobotatHandler" class is my moddified version of the Qualisis class that works with my Optitrack System.

`

    global event
    global Robotat
    event = Event()
    Robotat = RobotatHandler(RigidBodyID, event,\
                dataTime_Seconds, HOST, PORT,\
                send_full_pose)
    cflib.crtp.init_drivers()
    with SyncCrazyflie(uri, cf=Crazyflie(rw_cache='./cache')) as scf:
        cf = scf.cf
        Robotat.cf = cf
        activate_kalman_estimator(cf)
        activate_high_level_commander(cf)
        reset_estimator(cf)
        #activate_mellinger_controller(cf)

        cf.commander.set_client_xmode(True)
        cf.commander.send_setpoint(0, 0, 0, 0)

        for iter in range(0,3):
            print(f"Start in {3-iter:d} seconds")
            event.wait(1)

        log_config = estimatorLogData(cf)
        cf.commander.send_setpoint(0, 0, 0, 0)
        localHeight = 0.5
        try:
            while not keyboard.is_pressed('space') and not keyboard.is_pressed('enter'): 
                print("HOVERING...")
                it=0

                cf.param.set_value("kalman.quadIsFlying",1)

                while not keyboard.is_pressed('space'):
                    cf.commander.send_zdistance_setpoint(0,0,0,localHeight)
                    # cf.commander.send_hover_setpoint(0,0,0.5,0)                        
                    # cf.commander.send_position_setpoint(0, 0.5, 0.5, 0)
                    if it > 700:
                        cf.param.set_value("flightmode.poshold",1)
                        pass
                    event.wait(0.01)

                cf.param.set_value("kalman.quadIsFlying",0)
                print("\nClosing Log")
                log_config.stop()
                print("\nLanding and closing connection")
                if not keyboard.is_pressed('enter'):
                    landing(cf)
                event.wait(0.1)

        except BaseException as theError:
            # Don't execute the control code
            print('--------ERROR ON LOOP -----------') 
            print(f'Leaving because of ')
            print(theError)

`

**Comment by krichardsson** [comment]
Sending setpoints is a bit tricky, a good starting point is one of the examples scripts that we know is working to remove unknowns. I would suggest that you continue with the qualisys script, it will fly a figure 8 and make sure that works. When you know the system is working you can start to tweak stuff.
